# Data preparation

#### Read counts
This is the document containing the number of sequencing reads from each sample have been mapped to each MAG. Note that this is the raw data that needs to be further processed before running any statistics on them.

```{r load_count, message=FALSE}
read_counts <- read.csv("data/squirrels_counts.csv", sep=";") %>%
  rename(genome = 1)
```

Generate a vector of genome names to be employed for filtering and sorting data in downstream steps.
```{r genome_list, message=FALSE}
genomes <- read_counts$genome # create list of genome names
```

#### Genome coverage
Proportion of each genome covered by sequencing reads.

```{r genome_coverage, message=FALSE}
genome_coverage <- read.csv("data/squirrels_coverage.csv", sep=";") %>%
  rename(genome = 1) 
```

#### Sample metadata

```{r sample_metadata, message=FALSE}
sample_metadata <- read.csv("data/squirrels_metadata.csv",sep=";",header=T)
sample_metadata <- sample_metadata %>%  filter(!is.na(sampleID))
```

#### Genome metadata

```{r genome_metadata, message=FALSE}
genome_metadata <- read.csv("data/squirrels_mag_info.csv",sep=";") %>%
    rename(length=mag_size) %>%
    arrange(match(genome, read_counts$genome))
```

#### Coverage filtering
Genomes that have less than 30% of their length covered by reads are turned into zeros to account for the random allocation of reads across genomes due to mapping heuristics. 

```{r filter_coverage}
min_coverage=0.3
read_counts_filt <- genome_coverage %>%
  mutate(across(where(is.numeric), ~ ifelse(. > min_coverage, 1, 0))) %>%
  mutate(across(-1, ~ . * read_counts[[cur_column()]])) 
```

#### Generate genome count table
After filtering the low-coverage reads, read counts are transformed into genome counts using genome-length and read-length information.

```{r calc_genometable}
readlength=150 #change if sequencing read length is different
genome_counts <- read_counts_filt %>%
  mutate(across(where(is.numeric), ~ . / (genome_metadata$length / readlength) )) %>%
  select_if(~!all(. == 0)) #remove samples with no MAGs
```

#### Genome tree

```{r load_tree, message=FALSE, warning=FALSE}
genome_tree <- read.tree("data/squirrels.tree")
```

#### MAG functional annotations
These are the raw annotation table generated by DRAM, which are used to generate GIFT data using either distillR or KEGG.
```{r genome_annotations, message=FALSE}
genome_gifts <- read.csv("data/squirrels_distillr.csv", sep=";") 

genome_kegg <- read.csv("data/squirrels_merged_kegg.csv", sep=";")
```

#### Colour palettes
```{r colour_palettes}

ehi_phylum_colors <- read.table("data/ehi_phylum_colors.tsv",sep="\t",header=T,comment.char = "")
# Arrange colors alphabetically
# colors_alphabetic <- ehi_phylum_colors %>%
#   right_join(genome_metadata, by=join_by(phylum == phylum)) %>%
#   arrange(match(genome, genome_tree$tip.label)) %>%
#   select(phylum, colors) %>%
#   unique() %>%
#   arrange(phylum) %>%
#   select(colors) %>%
#   pull()

ehi_phylum_colors <- ehi_phylum_colors %>%
  mutate(colors2 = if_else(phylum=="p__Bacillota", '#226d1b', 
          if_else(phylum=="p__Bacillota_A", '#94cc84',
          if_else(phylum=="p__Bacillota_B", '#e36910',
          if_else(phylum=="p__Bacillota_C", '#f1a144',
          if_else(phylum=="p__Verrucomicrobiota", '#9d8dad',
          if_else(phylum=="p__Cyanobacteriota", '#088CA2', colors #'#08a26b'
                  )))))))
#ccffbc
#5c9c4f



phylum_colors <- ehi_phylum_colors %>% pull(colors2) 
names(phylum_colors) <- ehi_phylum_colors$phylum

squirrel_colors <- c("#999999", "#cc3333")
names(squirrel_colors) <- c("Sciurus carolinensis", "Sciurus vulgaris")

```

#### Wrap working objects
In the last step, the objects that are needed for downstream analyses are stored in an R object.

```{r wrap_objects}
save(read_counts, read_counts_filt, genome_counts, genome_tree, genome_metadata, genome_gifts, genome_kegg, sample_metadata, ehi_phylum_colors, phylum_colors, squirrel_colors, file = "data/squirrels_data.Rdata")
```
