# Data preparation

#### Read counts
This is the document containing the number of sequencing reads from each sample have been mapped to each MAG. Note that this is the raw data that needs to be further processed before running any statistics on them.

```{r load_count, message=FALSE}
read_counts <- read.csv("data/squirrels_counts.csv", sep=";") %>%
  rename(genome = 1)
```

Generate a vector of genome names to be employed for filtering and sorting data in downstream steps.
```{r genome_list, message=FALSE}
genomes <- read_counts$genome # create list of genome names
```

#### Genome coverage
Proportion of each genome covered by sequencing reads.

```{r genome_coverage, message=FALSE}
genome_coverage <- read.csv("data/squirrels_coverage.csv", sep=";") %>%
  rename(genome = 1)
```

#### Sample metadata

```{r sample_metadata, message=FALSE}
sample_metadata <- read.csv("data/squirrels_metadata.csv",sep=";",header=T)
sample_metadata <- sample_metadata %>%  filter(!is.na(sampleID))
```

#### Genome metadata

```{r genome_metadata, message=FALSE}
genome_metadata <- read.csv("data/squirrels_mag_info.csv",sep=";") %>%
    rename(length=mag_size) %>%
    arrange(match(genome, read_counts$genome))
```

#### Coverage filtering
Genomes that have less than 30% of their length covered by reads are turned into zeros to account for the random allocation of reads across genomes due to mapping heuristics.

```{r filter_coverage}
min_coverage=0.3
read_counts_filt <- genome_coverage %>%
  mutate(across(where(is.numeric), ~ ifelse(. > min_coverage, 1, 0))) %>%
  mutate(across(-1, ~ . * read_counts[[cur_column()]]))
```

#### Generate genome count table
After filtering the low-coverage reads, read counts are transformed into genome counts using genome-length and read-length information.

```{r calc_genometable}
readlength=150 #change if sequencing read length is different
genome_counts <- read_counts_filt %>%
  mutate(across(where(is.numeric), ~ . / (genome_metadata$length / readlength) )) %>%
  select_if(~!all(. == 0)) #remove samples with no MAGs
```

#### Genome tree

```{r load_tree, message=FALSE, warning=FALSE}
genome_tree <- read.tree("data/squirrels.tree")
```

#### Genome annotations
Downloading individual annotation files from ERDA using information in Airtable and writing them to a single compressed table takes a while. The following chunk only needs to be run once, to generate the ***genome_annotations*** table that is saved in the data directory. Note that the airtable connection requires a personal access token.

```{r download_genome_annotations, warning=FALSE, comments="", message=FALSE, eval=FALSE}
airtable("MAGs", "appWbHBNLE6iAsMRV") %>% #get base ID from Airtable browser URL
  read_airtable(., fields = c("ID","mag_name","number_genes","anno_url"), id_to_col = TRUE) %>% #get 3 columns from MAGs table
  filter(mag_name %in% paste0(genome_metadata$genome,".fa")) %>% #filter by MAG name
  filter(number_genes > 0) %>% #genes need to exist
  select(anno_url) %>% #list MAG annotation urls
  pull() %>%
  read_tsv() %>% #load all tables
  rename(gene=1, genome=2, contig=3) %>% #rename first 3 columns
  write_tsv(file="data/genome_annotations.tsv.xz") #write to overall compressed file
```

```{r load_genome_annotations, warning=FALSE, comments="", message=FALSE, eval=FALSE}
genome_annotations <- read_tsv("data/genome_annotations.tsv.xz") %>%
    rename(gene=1, genome=2, contig=3)
```

### Distill annotations into GIFTs

```{r distill_annotations, warning=FALSE, comments="", message=FALSE, eval=FALSE}
genome_gifts <- distill(genome_annotations,GIFT_db,genomecol=2,annotcol=c(9,10,19), verbosity=F) 

```

#### Colour palettes
```{r colour_palettes}

ehi_phylum_colors <- read.table("data/ehi_phylum_colors.tsv",sep="\t",header=T,comment.char = "")

phylum_colors <- ehi_phylum_colors %>%
  mutate(colors2 = if_else(phylum=="p__Bacillota", '#074759',
          if_else(phylum=="p__Bacillota_A", '#0c80a2',
          if_else(phylum=="p__Bacillota_B", '#10acd9',
          if_else(phylum=="p__Bacillota_C", '#7ddaf5',
          if_else(phylum=="p__Actinomycetota", '#C7B3E0',
          if_else(phylum=="p__Pseudomonadota", '#f5aa74',
          if_else(phylum=="p__Thermoplasmatota", '#f07e2a',
          if_else(phylum=="p__Verrucomicrobiota", '#77638A',
          if_else(phylum=="p__Cyanobacteriota", '#94cc84',
          if_else(phylum=="p__Elusimicrobiota", '#226d1b', colors
                  )))))))))))
#9E8DAD
#08D1D1
#ccffbc
#5c9c4f
#08a26b

ehi_colors <- phylum_colors %>% pull(colors)
names(ehi_colors) <- phylum_colors$phylum

custom_colors <- phylum_colors %>% pull(colors2)
names(custom_colors) <- phylum_colors$phylum

gift_colors <- read.table("data/gift_colors.tsv",sep="\t",header=T,comment.char = "")

squirrel_colors <- c("#999999", "#cc3333")
names(squirrel_colors) <- c("Sciurus carolinensis", "Sciurus vulgaris")

season_colors <- c("#76b183","#d98a41","#6b7398") #e5bd5b
names(season_colors) <- c("spring-summer", "autumn", "winter")

area_colors <- c("#3d6a4b","#94a766","#c1cca7")
names(area_colors) <- c("rural", "suburban", "urban")

sex_colors <- c("turquoise3", "indianred2")
names(sex_colors) <- c("M", "F")

```

#### Wrap working objects
In the last step, the objects that are needed for downstream analyses are stored in an R object.

```{r wrap_objects}
save(read_counts, read_counts_filt, genome_counts, genome_tree, genome_metadata, genome_gifts, sample_metadata, custom_colors, ehi_colors, phylum_colors, gift_colors, squirrel_colors, season_colors, area_colors, sex_colors, file = "data/squirrels_data.Rdata")
```
