---
title: "Diversity analysis"
author: "Claudia Romeo"
date: "2024-02-20"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

##Load libraries
```{r load_libraries, include=FALSE}
library(tidyverse)
library(dplyr)
library(ape)
library(spaa)
library(vegan)
library(Rtsne)
library(xfun)
library(knitr)
library(ggplot2)
library(ggrepel)
library(ggpubr)
library(ggnewscale)
library(gridExtra)


#install_github("anttonalberdi/hilldiv2")
library(hilldiv2)

```



## Load data

```{r load_datasets, include=FALSE}

load("data/squirrels_data.Rdata")
```


## Data preparation

```{r data_preparation, message=FALSE}

#Change genome names column to row names
genome_counts <- genome_counts %>%
  column_to_rownames(var="genome")
genome_kegg <- genome_kegg %>%
  column_to_rownames(var="genome")
genome_gifts <- genome_gifts %>%
  column_to_rownames(var="genome")

#Get list of present MAGs
present_MAGs <- genome_counts %>%
  filter(rowSums(.[, -1]) != 0) %>%
  rownames()

#Remove samples with all zeros (no data after filtering)
genome_counts_filt <- genome_counts %>%
  select_if(~!all(. == 0))


#Align KEGG and distillr annotations with present MAGs and remove all-zero and all-one traits
present_MAGs <- present_MAGs[present_MAGs %in% rownames(genome_kegg)]
genome_kegg_filt <- genome_kegg[present_MAGs,] %>%
  select_if(~!all(. == 0)) %>%  #remove all-zero modules
  select_if(~!all(. == 1)) #remove all-one modules

genome_gifts_filt <- genome_gifts[present_MAGs,] %>%
  select_if(~!all(. == 0)) %>%  #remove all-zero modules
  select_if(~!all(. == 1)) #remove all-one modules

#Align tree with present MAGs
tree_filt <- keep.tip(genome_tree,present_MAGs)

#Filter count table to only contain present MAGs after KEGG filtering
genome_counts_filt <- genome_counts[present_MAGs,]

#Calculate sequence fractions for each samples
sequence_fractions <- read_counts %>%
  pivot_longer(-genome, names_to = "sample", values_to = "value") %>%
  group_by(sample) %>%
  summarise(mags = sum(value)) %>%
  left_join(sample_metadata, by = join_by(sample == sample))  %>%
  select(sample,mags,metagenomic_bases,host_bases,bases_lost_fastp_percent) %>%
  mutate(mags_bases = mags*146) %>%
  mutate(lowqual_bases = ((metagenomic_bases+host_bases)/(1-bases_lost_fastp_percent))-(metagenomic_bases+host_bases)) %>%
  mutate(unmapped_bases = metagenomic_bases - mags_bases) %>%
  mutate(unmapped_bases = ifelse(unmapped_bases < 0, 0, unmapped_bases)) %>%
  select(sample,mags_bases,unmapped_bases,host_bases,lowqual_bases)

```

## Diversity metrics calculation

```{r alpha_diversity}
#ALPHA DIVERSITY
q0n <- hilldiv2::hilldiv(genome_counts,q=0) %>% c()
q1n <- hilldiv2::hilldiv(genome_counts,q=1) %>% c()
q1p <- hilldiv2::hilldiv(genome_counts,q=1,tree=genome_tree) %>% c()
dist <- hilldiv2::traits2dist(genome_kegg_filt, method="gower")
q1f <- hilldiv2::hilldiv(genome_counts_filt,q=1,dist=dist) %>% c()
dist2 <- hilldiv2::traits2dist(genome_gifts_filt, method="gower")
q1f2 <- hilldiv2::hilldiv(genome_counts_filt,q=1,dist=dist2) %>% c()

```


```{r beta_diversity, eval=FALSE}
#BETA DIVERSITY
#neutral beta div
beta_q1n <-hilldiv2::hillpair(genome_counts, q=1, metric="S")
#phylogenetic beta div (ERRORS)
#beta_q1p <-hilldiv2::hillpair(count_table_cov_size, q=1, tree=tree, metric="S")
#functional beta div
beta_q1n_fun <-hilldiv2::hillpair(genome_counts_filt, q=1, dist=dist)
beta_q1n_fun2 <-hilldiv2::hillpair(genome_counts_filt, q=1, dist=dist2)
```

## Alpha Diversity summary

```{r}
# Merge all metrics
alpha_div <- cbind(sample=colnames(genome_counts),richness=q0n,neutral=round(q1n,3),phylo=round(q1p,3),func=round(q1f,3),func2=round(q1f2,3)) %>%
  as.data.frame()
columns <- c("richness","neutral","func","func2","phylo", "mapped","total")

# Add amount of sequencing data to the table
alpha_div <- alpha_div %>%
  left_join(sequence_fractions, by = join_by(sample == sample)) %>% #add sequencing depth information
  mutate(mapped=round(mags_bases/1000000000,3)) %>% #modify depth to million reads
  mutate(total=round((mags_bases+unmapped_bases+host_bases+lowqual_bases)/1000000000,3)) %>%
  select(sample,richness,neutral,phylo,func,func2,mapped,total) %>%
  mutate(across(-1, as.numeric))

# Export alpha diversity metrics to be used later in script 05-diversity_models
write.table(alpha_div, file = "data/alpha_div.csv", row.names = FALSE, dec = ".", sep = ";", 
              quote = FALSE)

alpha_div %>%
  pivot_longer(-sample, names_to = "data", values_to = "value") %>%
  mutate(data = factor(data, levels = columns))   %>%
  ggplot(aes(x=value, y=sample)) +
  geom_bar(stat='identity', fill="#6c9ebc") +
  facet_wrap(~data,  scales="free_x", ncol=6) +
  theme_classic() +
  theme(
    strip.background = element_blank(),
    panel.grid.minor.x = element_line( size=.1, color="grey" ),
    axis.title.x = element_blank(),
    axis.title.y = element_blank(),
    axis.text.x = element_text(angle = 45, hjust = 1)
  )

#table
kable(alpha_div)

```

## Alpha Diversity comparisons

# By species

```{r}
squirrel_colors <- c("#999999", "#cc3333")

#neutral alpha
alpha_div %>%
  select(sample,neutral) %>%
  pivot_longer(-sample, names_to = "data", values_to = "value") %>%
  mutate(data = factor(data, levels = columns))   %>%
  left_join(sample_metadata, by = join_by(sample == sample)) %>%
  #mutate(location=paste0(round(longitude,2),"_",round(latitude,2))) %>%
  ggboxplot(., x = "species", y = "value", color = "species", fill="white", add="jitter") +
      scale_color_manual(values=squirrel_colors) +
      scale_fill_manual(values=paste0(squirrel_colors)) +
      stat_compare_means() +
      theme_classic() +
      labs(y = "Neutral Hill numbers") +
      theme(
        legend.position = "none",
        axis.title.x = element_blank())
```

```{r}
#phylogenetic alpha
alpha_div %>%
  select(sample,phylo) %>%
  pivot_longer(-sample, names_to = "data", values_to = "value") %>%
  mutate(data = factor(data, levels = columns))   %>%
  left_join(sample_metadata, by = join_by(sample == sample)) %>%
  mutate(location=paste0(round(longitude,2),"_",round(latitude,2))) %>%
  ggboxplot(., x = "species", y = "value", color = "species", fill="white", add="jitter") +
  scale_color_manual(values=squirrel_colors) +
  scale_fill_manual(values=paste0(squirrel_colors)) +
  stat_compare_means() +
  theme_classic() +
  labs(y = "Phylogenetic Hill numbers") +
  theme(
    legend.position = "none",
    axis.title.x = element_blank())

```


```{r}

#functional alpha (kegg-based)
alpha_div %>%
  select(sample,func) %>%
  pivot_longer(-sample, names_to = "data", values_to = "value") %>%
  mutate(data = factor(data, levels = columns))   %>%
  left_join(sample_metadata, by = join_by(sample == sample)) %>%
  mutate(location=paste0(round(longitude,2),"_",round(latitude,2))) %>%
  ggboxplot(., x = "species", y = "value", color = "species", fill="white", add="jitter") +
  scale_color_manual(values=squirrel_colors) +
  scale_fill_manual(values=paste0(squirrel_colors)) +
  stat_compare_means() +
  theme_classic() +
  labs(y = "Functional Hill numbers") +
  theme(
    legend.position = "none",
    axis.title.x = element_blank())

```

```{r}

#functional alpha (distillr-based)
alpha_div %>%
  select(sample,func2) %>%
  pivot_longer(-sample, names_to = "data", values_to = "value") %>%
  mutate(data = factor(data, levels = columns))   %>%
  left_join(sample_metadata, by = join_by(sample == sample)) %>%
  mutate(location=paste0(round(longitude,2),"_",round(latitude,2))) %>%
  ggboxplot(., x = "species", y = "value", color = "species", fill="white", add="jitter") +
  scale_color_manual(values=squirrel_colors) +
  scale_fill_manual(values=paste0(squirrel_colors)) +
  stat_compare_means() +
  theme_classic() +
  labs(y = "Functional Hill numbers") +
  theme(
    legend.position = "none",
    axis.title.x = element_blank())

```

# By species and area type

```{r}
sample_metadata$area_type <-factor(sample_metadata$area_type, levels = c("rural", "suburban", "urban"))
area_colors <- c("#76b183","#d57d2c","#6b7398")

#neutral alpha by species*area_type
alpha_div %>%
  select(sample,neutral) %>%
  pivot_longer(-sample, names_to = "data", values_to = "value") %>%
  mutate(data = factor(data, levels = columns))   %>%
  left_join(sample_metadata, by = join_by(sample == sample)) %>%
  ggboxplot(., x = "species", y = "value", color = "area_type", fill="white", add="jitter") +
  scale_color_manual(values=area_colors) +
  scale_fill_manual(values=paste0(area_colors)) +
  stat_compare_means() +
  theme_classic() +
  labs(y = "Neutral Hill numbers") +
  theme(
    legend.position = "top",
    legend.box = "horizontal",
    axis.title.x = element_blank()) +
  guides(color=guide_legend(title="Area type"), fill="none")


```

```{r}
#phylogenetic alpha by species*area_type
alpha_div %>%
  select(sample,phylo) %>%
  pivot_longer(-sample, names_to = "data", values_to = "value") %>%
  mutate(data = factor(data, levels = columns))   %>%
  left_join(sample_metadata, by = join_by(sample == sample)) %>%
  ggboxplot(., x = "species", y = "value", color = "area_type", fill="white", add="jitter") +
  scale_color_manual(values=area_colors) +
  scale_fill_manual(values=paste0(area_colors)) +
  stat_compare_means() +
  theme_classic() +
  labs(y = "Phylogenetic Hill numbers") +
  theme(
    legend.position = "top",
    legend.box = "horizontal",
    axis.title.x = element_blank()) +
  guides(color=guide_legend(title="Area type"), fill="none")
```

```{r}
#functional alpha (distillr-based) by species*area_type
alpha_div %>%
  select(sample,func2) %>%
  pivot_longer(-sample, names_to = "data", values_to = "value") %>%
  mutate(data = factor(data, levels = columns))   %>%
  left_join(sample_metadata, by = join_by(sample == sample)) %>%
  ggboxplot(., x = "species", y = "value", color = "area_type", fill="white", add="jitter") +
  scale_color_manual(values=area_colors) +
  scale_fill_manual(values=paste0(area_colors)) +
  stat_compare_means() +
  theme_classic() +
  labs(y = "Functional Hill numbers") +
  theme(
    legend.position = "top",
    legend.box = "horizontal",
    axis.title.x = element_blank()) +
  guides(color=guide_legend(title="Area type"), fill="none")
```
# By species and season

```{r}
sample_metadata$season <-factor(sample_metadata$season, levels = c("spring-summer", "autumn", "winter"))
season_colors <- c("#76b183","#e5bd5b","#6b7398")

#neutral alpha by species*season
alpha_div %>%
  select(sample,neutral) %>%
  pivot_longer(-sample, names_to = "data", values_to = "value") %>%
  mutate(data = factor(data, levels = columns))   %>%
  left_join(sample_metadata, by = join_by(sample == sample)) %>%
  ggboxplot(., x = "species", y = "value", color = "season", fill="white", add="jitter") +
  scale_color_manual(values=season_colors) +
  scale_fill_manual(values=paste0(season_colors)) +
  stat_compare_means() +
  theme_classic() +
  labs(y = "Neutral Hill numbers") +
  theme(
    legend.position = "top",
    legend.box = "horizontal",
    axis.title.x = element_blank()) +
  guides(color=guide_legend(title="Season"), fill="none")

```

```{r}

#phylogenetic alpha by species*season
alpha_div %>%
  select(sample,phylo) %>%
  pivot_longer(-sample, names_to = "data", values_to = "value") %>%
  mutate(data = factor(data, levels = columns))   %>%
  left_join(sample_metadata, by = join_by(sample == sample)) %>%
  ggboxplot(., x = "species", y = "value", color = "season", fill="white", add="jitter") +
  scale_color_manual(values=season_colors) +
  scale_fill_manual(values=paste0(season_colors)) +
  stat_compare_means() +
  theme_classic() +
  labs(y = "Phylogenetic Hill numbers") +
  theme(
    legend.position = "top",
    legend.box = "horizontal",
    axis.title.x = element_blank()) +
  guides(color=guide_legend(title="Season"), fill="none")

```

```{r}

#functional alpha (distillr-based) by species*season
alpha_div %>%
  select(sample,func2) %>%
  pivot_longer(-sample, names_to = "data", values_to = "value") %>%
  mutate(data = factor(data, levels = columns))   %>%
  left_join(sample_metadata, by = join_by(sample == sample)) %>%
  ggboxplot(., x = "species", y = "value", color = "season", fill="white", add="jitter") +
  scale_color_manual(values=season_colors) +
  scale_fill_manual(values=paste0(season_colors)) +
  stat_compare_means() +
  theme_classic() +
  labs(y = "Functional Hill numbers") +
  theme(
    legend.position = "top",
    legend.box = "horizontal",
    axis.title.x = element_blank()) +
  guides(color=guide_legend(title="Season"), fill="none")


```

# Alpha diversity and sequencing effort

```{r}

#sequencing effort and diversity
ggplot(alpha_div, aes(x=mapped,y=neutral,label=sample)) +
  geom_smooth(method='lm', formula= y~x, color='#e08dde', fill='#e08dde') +
  geom_point(alpha=0.5, color="#6c9ebc") +
  geom_label_repel(max.overlaps = 100, cex=0.7) +
  labs(x = "GBs mapped to MAGs", y = "Neutral diversity (effective number of MAGs)") +
  theme_classic() +
  theme(legend.position="none")
```


## Beta Diversity comparisons

# NMDS on Neutral Beta Diversity

```{r NMDS_beta, eval=FALSE}

#neutral beta diversity NMDS
sample_metadata_adonis <- sample_metadata %>%
  filter(sample %in% labels(beta_q1n)) %>%
  arrange(sample) %>%
  #mutate(location=paste0(round(longitude,2),"_",round(latitude,2))) %>%
  select(sample,species,sex,area_type,macroarea,season) %>%
  select_if(~ length(unique(.)) > 1) %>% #remove columns with all-identical values
  column_to_rownames(var = "sample") %>%
  as.data.frame()

adonis2(formula=beta_q1n ~ ., data=sample_metadata_adonis[labels(beta_q1n),], permutations=999) %>%
  as.matrix() %>%
  print()

```


```{r , eval=FALSE}
beta_q1n_nmds <- beta_q1n %>%
  metaMDS(.,trymax = 500, k=2, verbosity=FALSE) %>%
  vegan::scores() %>%
  as_tibble(., rownames = "sample") %>%
  left_join(sample_metadata, by = join_by(sample == sample))

beta_q1n_nmds %>%
  group_by(species) %>%
  filter(sample !="EHI00420") %>% #remove outlier
  mutate(x_cen = mean(NMDS1, na.rm = TRUE)) %>%
  mutate(y_cen = mean(NMDS2, na.rm = TRUE)) %>%
  ungroup() %>%
  ggplot(., aes(x=NMDS1,y=NMDS2, color=species, shape=area_type)) +
  scale_color_manual(values=squirrel_colors) +
  geom_point(size=2.5) +
  geom_segment(aes(x=x_cen, y=y_cen, xend=NMDS1, yend=NMDS2), alpha=0.2) +
  theme_classic() +
  theme(legend.position="right", legend.box="vertical") +
  guides(color=guide_legend(title="Species"), shape=guide_legend(title="Area type"))
```

# NMDS on Functional Beta Diversity

```{r , eval=FALSE}
#functional (kegg-based) beta diversity NMDS
sample_table_adonis_fun <- sample_metadata %>%
  filter(sample %in% labels(beta_q1n_fun$S)) %>%
  arrange(sample) %>%
  #mutate(location=paste0(round(longitude,2),"_",round(latitude,2))) %>%
  select(sample,species,sex,area_type,macroarea,season) %>%
  select_if(~ length(unique(.)) > 1) %>% #remove columns with all-identical values
  column_to_rownames(var = "sample") %>%
  as.data.frame()

adonis2(formula=beta_q1n_fun$S ~ ., data=sample_table_adonis_fun[labels(beta_q1n_fun$S),], permutations=999) %>%
  as.matrix() %>%
  print()
```

```{r , eval=FALSE}
beta_q1n_fun_nmds <- beta_q1n_fun$S %>%
  metaMDS(.,trymax = 500, k=2, verbosity=FALSE) %>%
  vegan::scores() %>%
  as_tibble(., rownames = "sample") %>%
  left_join(sample_metadata, by = join_by(sample == sample))

beta_q1n_fun_nmds %>%
  group_by(species) %>%
  filter(sample !="EHI00420") %>% #remove outlier
  mutate(x_cen = mean(NMDS1, na.rm = TRUE)) %>%
  mutate(y_cen = mean(NMDS2, na.rm = TRUE)) %>%
  ungroup() %>%
  ggplot(., aes(x=NMDS1,y=NMDS2, color=species, shape=area_type)) +
  scale_color_manual(values=squirrel_colors) +
  geom_point(size=2.5) +
  geom_segment(aes(x=x_cen, y=y_cen, xend=NMDS1, yend=NMDS2), alpha=0.2) +
  theme_classic() +
  theme(legend.position="right", legend.box="vertical") +
  guides(color=guide_legend(title="Species"), shape=guide_legend(title="Area type"))

```

```{r , eval=FALSE}
#functional (distillr-based) beta diversity NMDS
sample_table_adonis_fun2 <- sample_metadata %>%
  filter(sample %in% labels(beta_q1n_fun2$S)) %>%
  arrange(sample) %>%
  #mutate(location=paste0(round(longitude,2),"_",round(latitude,2))) %>%
  select(sample,species,sex,area_type,macroarea,season) %>%
  select_if(~ length(unique(.)) > 1) %>% #remove columns with all-identical values
  column_to_rownames(var = "sample") %>%
  as.data.frame()

adonis2(formula=beta_q1n_fun2$S ~ ., data=sample_table_adonis_fun2[labels(beta_q1n_fun2$S),], permutations=999) %>%
  as.matrix() %>%
  print()
```

```{r , eval=FALSE}
beta_q1n_fun2_nmds <- beta_q1n_fun2$S %>%
  metaMDS(.,trymax = 500, k=2, verbosity=FALSE) %>%
  vegan::scores() %>%
  as_tibble(., rownames = "sample") %>%
  left_join(sample_metadata, by = join_by(sample == sample))

beta_q1n_fun2_nmds %>%
  group_by(species) %>%
  filter(sample !="EHI00420") %>% #remove outlier
  mutate(x_cen = mean(NMDS1, na.rm = TRUE)) %>%
  mutate(y_cen = mean(NMDS2, na.rm = TRUE)) %>%
  ungroup() %>%
  ggplot(., aes(x=NMDS1,y=NMDS2, color=species, shape=area_type)) +
  scale_color_manual(values=squirrel_colors) +
  geom_point(size=2.5) +
  geom_segment(aes(x=x_cen, y=y_cen, xend=NMDS1, yend=NMDS2), alpha=0.2) +
  theme_classic() +
  theme(legend.position="right", legend.box="vertical") +
  guides(color=guide_legend(title="Species"), shape=guide_legend(title="Area type"))

```