---
title: "Diversity analysis"
author: "Claudia Romeo"
date: "2024-02-20"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load libraries
```{r load_libraries, include=FALSE}
library(tidyverse)
library(dplyr)
library(ape)
library(spaa)
library(vegan)
library(Rtsne)
library(xfun)
library(knitr)
library(ggplot2)
library(ggrepel)
library(ggpubr)
library(ggnewscale)
library(gridExtra)
library(tinytable)
library(lme4)
library(emmeans)
library(effects)
library(car)
library(lmerTest)
library(ggResidpanel)
library(gtsummary)
library(broom.mixed)

#install_github("anttonalberdi/hilldiv2")
library(hilldiv2)

```


# Load data and set environment

```{r load_datasets, include=FALSE}

rm(list=ls()) #clear environment

load("data/squirrels_data.Rdata")

singlem <- read.csv("data/singlem.csv",sep=";",header=T)

options(contrasts = c('contr.sum','contr.poly'))

```


# Data preparation

```{r data_preparation, message=FALSE}

#Change genome names column to row names
genome_counts <- genome_counts %>%
  column_to_rownames(var="genome")
genome_gifts <- genome_gifts %>%
  column_to_rownames(var="genome")

#Get list of present MAGs
present_MAGs <- genome_counts %>%
  filter(rowSums(.[, -1]) != 0) %>%
  rownames()

#Remove samples with all zeros (no data after filtering)
genome_counts_filt <- genome_counts %>%
  select_if(~!all(. == 0))

#Align distillr annotations with present MAGs and remove all-zero and all-one traits
present_MAGs <- present_MAGs[present_MAGs %in% rownames(genome_gifts)]

genome_gifts_filt <- genome_gifts[present_MAGs,] %>%
  select_if(~!all(. == 0)) %>%  #remove all-zero modules
  select_if(~!all(. == 1)) #remove all-one modules

#Align tree with present MAGs
tree_filt <- keep.tip(genome_tree,present_MAGs)

#Filter count table to only contain present MAGs after gifts filtering
genome_counts_filt <- genome_counts[present_MAGs,]

#Calculate sequence fractions for each samples
sequence_fractions <- read_counts %>%
  pivot_longer(-genome, names_to = "sample", values_to = "value") %>%
  group_by(sample) %>%
  summarise(mags = sum(value)) %>%
  left_join(sample_metadata, by = join_by(sample == sample))  %>%
  select(sample,mags,metagenomic_bases,host_bases,bases_lost_fastp_percent) %>%
  mutate(mags_bases = mags*146) %>%
  mutate(lowqual_bases = ((metagenomic_bases+host_bases)/(1-bases_lost_fastp_percent))-(metagenomic_bases+host_bases)) %>%
  mutate(unmapped_bases = metagenomic_bases - mags_bases) %>%
  mutate(unmapped_bases = ifelse(unmapped_bases < 0, 0, unmapped_bases)) %>%
  select(sample,mags_bases,unmapped_bases,host_bases,lowqual_bases)


```

# Alpha Diversity metrics 

```{r alpha_diversity}
#ALPHA DIVERSITY
q0n <- hilldiv2::hilldiv(genome_counts,q=0) %>% c()
q1n <- hilldiv2::hilldiv(genome_counts,q=1) %>% c()
q1p <- hilldiv2::hilldiv(genome_counts,q=1,tree=genome_tree) %>% c()
dist <- hilldiv2::traits2dist(genome_gifts_filt, method="gower")
q1f <- hilldiv2::hilldiv(genome_counts_filt,q=1,dist=dist) %>% c()

```

## Alpha Diversity summary

```{r alpha_summary, warning=FALSE}

# Merge all metrics
alpha_div <- cbind(sample=colnames(genome_counts),richness=q0n,neutral=round(q1n,3),phylo=round(q1p,3),func=round(q1f,3)) %>%
  as.data.frame()
columns <- c("richness","neutral","phylo","func", "mapped","total")

# Add amount of sequencing data to the table
alpha_div <- alpha_div %>%
  left_join(sequence_fractions, by = join_by(sample == sample)) %>% #add sequencing depth information
  mutate(mapped=round(mags_bases/1000000000,3)) %>% #modify depth to million reads
  mutate(total=round((mags_bases+unmapped_bases+host_bases+lowqual_bases)/1000000000,3)) %>%
  select(sample,richness,neutral,phylo,func,mapped,total) %>%
  mutate(across(-1, as.numeric))

squirrel_colors <- c("#999999", "#cc3333")

alpha_div %>%
  left_join(sample_metadata, by='sample') %>%
  select(sample, species, richness, neutral,phylo,func, mapped, total) %>%
  mutate(species = factor(species), # Convert to factor if necessary
        sample = factor(sample, levels = unique(sample)[order(species)])) %>%
  pivot_longer(-c(sample, species), names_to = "data", values_to = "value") %>%
  mutate(data = factor(data, levels = columns)) %>%
  ggplot(aes(x=value, y=sample, fill=species)) +
  geom_bar(stat='identity') +
  scale_fill_manual(values=squirrel_colors) +
  facet_wrap(~ data, scales="free_x", ncol=6) +
  #facet_grid(species ~ data, scales="free_x", space="fixed") +
  #force_panelsizes(ro#ws = 2, cols = 3, TRUE) +
  theme_classic() +
  theme(
    strip.background = element_blank(),
    panel.grid.minor.x = element_line(size=.1, color="grey"),
    panel.spacing = unit(0, "lines"),
    axis.title.x = element_blank(),
    axis.title.y = element_blank(),
    axis.text.x = element_text(angle = 45, hjust = 1),
    axis.text.y = element_blank(),
    legend.position = "none"
  )

# #table
# kable(alpha_div)

```

## Alpha Diversity comparisons

### By species and sex

```{r alpha_sex_plots, fig.dim = c(8, 10)}

squirrel_colors <- c("#999999", "#cc3333")

sex_colors <- c("turquoise3", "indianred2")


neutral.sex <- alpha_div %>%
            select(sample,neutral) %>%
            pivot_longer(-sample, names_to = "data", values_to = "value") %>%
            mutate(data = factor(data, levels = columns))   %>%
            left_join(sample_metadata, by = join_by(sample == sample)) %>%
            ggboxplot(., x = "species", y = "value", color = "sex", fill="white", add="jitter") +
                  scale_color_manual(values=sex_colors) +
                  scale_fill_manual(values=paste0(squirrel_colors)) +
                  stat_compare_means() +
                  theme_classic() +
                  labs(y = "Neutral Hill numbers") +
                  theme(
                      legend.position = "none",
                      axis.title.x = element_blank())

phylo.sex <- alpha_div %>%
            select(sample,phylo) %>%
            pivot_longer(-sample, names_to = "data", values_to = "value") %>%
            mutate(data = factor(data, levels = columns))   %>%
            left_join(sample_metadata, by = join_by(sample == sample)) %>%
            ggboxplot(., x = "species", y = "value", color = "sex", fill="white", add="jitter") +
                  scale_color_manual(values=sex_colors) +
                  scale_fill_manual(values=paste0(squirrel_colors)) +
                  stat_compare_means() +
                  theme_classic() +
                  labs(y = "Phylogenetic Hill numbers") +
                  theme(
                      legend.position = "none",
                      axis.title.x = element_blank())

func.sex <- alpha_div %>%
            select(sample,func) %>%
            pivot_longer(-sample, names_to = "data", values_to = "value") %>%
            mutate(data = factor(data, levels = columns))   %>%
            left_join(sample_metadata, by = join_by(sample == sample)) %>%
            ggboxplot(., x = "species", y = "value", color = "sex", fill="white", add="jitter") +
                  scale_color_manual(values=sex_colors) +
                  scale_fill_manual(values=paste0(squirrel_colors)) +
                  stat_compare_means() +
                  theme_classic() +
                  labs(y = "Functional Hill numbers") +
                  theme(
                      legend.position = "none",
                      axis.title.x = element_blank())


sex.legend <- get_legend(neutral.sex)


ggarrange(neutral.sex, phylo.sex, func.sex, #+ rremove("x.text"), 
          legend.grob = sex.legend, legend="right", common.legend = TRUE,
          #labels = c("A", "B", "C"),
          ncol = 1, nrow = 3)
```


### By species and urbanisation

```{r alpha_urb_plots, fig.dim = c(8, 10)}

sample_metadata$area_type <-factor(sample_metadata$area_type, levels = c("rural", "suburban", "urban"))
area_colors <- c("#76b183","#d57d2c","#6b7398")

#neutral alpha by species*area_type
neutral.urb <- alpha_div %>%
            select(sample,neutral) %>%
            pivot_longer(-sample, names_to = "data", values_to = "value") %>%
            mutate(data = factor(data, levels = columns))   %>%
            left_join(sample_metadata, by = join_by(sample == sample)) %>%
            ggboxplot(., x = "species", y = "value", color = "area_type", fill="white", add="jitter") +
            scale_color_manual(values=area_colors) +
            scale_fill_manual(values=paste0(area_colors)) +
            stat_compare_means() +
            theme_classic() +
            labs(y = "Neutral Hill numbers") +
            theme(
              legend.position = "none",
              axis.title.x = element_blank()) +
            guides(color=guide_legend(title="Urbanisation"), fill="none")

#phylogenetic alpha by species*area_type
phylo.urb <- alpha_div %>%
            select(sample,phylo) %>%
            pivot_longer(-sample, names_to = "data", values_to = "value") %>%
            mutate(data = factor(data, levels = columns))   %>%
            left_join(sample_metadata, by = join_by(sample == sample)) %>%
            ggboxplot(., x = "species", y = "value", color = "area_type", fill="white", add="jitter") +
                  scale_color_manual(values=area_colors) +
                  scale_fill_manual(values=paste0(area_colors)) +
                  stat_compare_means() +
                  theme_classic() +
                  labs(y = "Phylogenetic Hill numbers") +
                  theme(
                      legend.position = "none",
                      axis.title.x = element_blank())

#functional (distillr-based) alpha by species*area_type
func.urb <- alpha_div %>%
            select(sample,func) %>%
            pivot_longer(-sample, names_to = "data", values_to = "value") %>%
            mutate(data = factor(data, levels = columns))   %>%
            left_join(sample_metadata, by = join_by(sample == sample)) %>%
            ggboxplot(., x = "species", y = "value", color = "area_type", fill="white", add="jitter") +
                  scale_color_manual(values=area_colors) +
                  scale_fill_manual(values=paste0(area_colors)) +
                  stat_compare_means() +
                  theme_classic() +
                  labs(y = "Functional Hill numbers") +
                  theme(
                      legend.position = "none",
                      axis.title.x = element_blank())


urb.legend <- get_legend(neutral.urb)


ggarrange(neutral.urb, phylo.urb, func.urb, #+ rremove("x.text"), 
          legend.grob = urb.legend, legend="right", common.legend = TRUE,
          #labels = c("A", "B", "C"),
          ncol = 1, nrow = 3)
```



### By species and season

```{r alpha_season_plots, fig.dim = c(8, 10)}

sample_metadata$season <-factor(sample_metadata$season, levels = c("spring-summer", "autumn", "winter"))
season_colors <- c("#76b183","#e5bd5b","#6b7398")

#neutral alpha by species*season
neutral.seas <- alpha_div %>%
            select(sample,neutral) %>%
            pivot_longer(-sample, names_to = "data", values_to = "value") %>%
            mutate(data = factor(data, levels = columns))   %>%
            left_join(sample_metadata, by = join_by(sample == sample)) %>%
            ggboxplot(., x = "species", y = "value", color = "season", fill="white", add="jitter") +
            scale_color_manual(values=season_colors) +
            scale_fill_manual(values=paste0(season_colors)) +
            stat_compare_means() +
            theme_classic() +
            labs(y = "Neutral Hill numbers") +
            theme(
              legend.position = "none",
              axis.title.x = element_blank()) +
            guides(color=guide_legend(title="Season"), fill="none")

#phylogenetic alpha by species*season
phylo.seas <- alpha_div %>%
            select(sample,phylo) %>%
            pivot_longer(-sample, names_to = "data", values_to = "value") %>%
            mutate(data = factor(data, levels = columns))   %>%
            left_join(sample_metadata, by = join_by(sample == sample)) %>%
            ggboxplot(., x = "species", y = "value", color = "season", fill="white", add="jitter") +
                  scale_color_manual(values=season_colors) +
                  scale_fill_manual(values=paste0(season_colors)) +
                  stat_compare_means() +
                  theme_classic() +
                  labs(y = "Phylogenetic Hill numbers") +
                  theme(
                      legend.position = "none",
                      axis.title.x = element_blank())

#functional (distillr-based) alpha by species*season
func.seas <- alpha_div %>%
            select(sample,func) %>%
            pivot_longer(-sample, names_to = "data", values_to = "value") %>%
            mutate(data = factor(data, levels = columns))   %>%
            left_join(sample_metadata, by = join_by(sample == sample)) %>%
            ggboxplot(., x = "species", y = "value", color = "season", fill="white", add="jitter") +
                  scale_color_manual(values=season_colors) +
                  scale_fill_manual(values=paste0(season_colors)) +
                  stat_compare_means() +
                  theme_classic() +
                  labs(y = "Functional Hill numbers") +
                  theme(
                      legend.position = "none",
                      axis.title.x = element_blank())


seas.legend <- get_legend(neutral.seas)


ggarrange(neutral.seas, phylo.seas, func.seas, #+ rremove("x.text"), 
          legend.grob = seas.legend, legend="right", common.legend = TRUE,
          #labels = c("A", "B", "C"),
          ncol = 1, nrow = 3)

```


### By species and development

```{r alpha_dev_plots, fig.dim = c(8, 10)}

#neutral alpha by species*season
neutral.dev <- alpha_div %>%
            select(sample,neutral) %>%
            pivot_longer(-sample, names_to = "data", values_to = "value") %>%
            mutate(data = factor(data, levels = columns))   %>%
            left_join(sample_metadata, by = join_by(sample == sample)) %>%
            ggboxplot(., x = "species", y = "value", color = "development", fill="white", add="jitter") +
            stat_compare_means() +
            theme_classic() +
            labs(y = "Neutral Hill numbers") +
            theme(
              legend.position = "none",
              axis.title.x = element_blank()) +
            guides(color=guide_legend(title="Development"), fill="none")

#phylogenetic alpha by species*season
phylo.dev <- alpha_div %>%
            select(sample,phylo) %>%
            pivot_longer(-sample, names_to = "data", values_to = "value") %>%
            mutate(data = factor(data, levels = columns))   %>%
            left_join(sample_metadata, by = join_by(sample == sample)) %>%
            ggboxplot(., x = "species", y = "value", color = "development", fill="white", add="jitter") +
                  stat_compare_means() +
                  theme_classic() +
                  labs(y = "Phylogenetic Hill numbers") +
                  theme(
                      legend.position = "none",
                      axis.title.x = element_blank())

#functional (distillr-based) alpha by species*season
func.dev <- alpha_div %>%
            select(sample,func) %>%
            pivot_longer(-sample, names_to = "data", values_to = "value") %>%
            mutate(data = factor(data, levels = columns))   %>%
            left_join(sample_metadata, by = join_by(sample == sample)) %>%
            ggboxplot(., x = "species", y = "value", color = "development", fill="white", add="jitter") +
                  stat_compare_means() +
                  theme_classic() +
                  labs(y = "Functional Hill numbers") +
                  theme(
                      legend.position = "none",
                      axis.title.x = element_blank())


dev.legend <- get_legend(neutral.seas)


ggarrange(neutral.dev, phylo.dev, func.dev, #+ rremove("x.text"), 
          legend.grob = seas.legend, legend="right", common.legend = TRUE,
          #labels = c("A", "B", "C"),
          ncol = 1, nrow = 3)


```


### Alpha diversity and sequencing effort

```{r alpha_sequencing, message=FALSE, warning=FALSE}

#sequencing effort and diversity
ggplot(alpha_div, aes(x=mapped,y=neutral,label=sample)) +
  geom_smooth(method='lm', formula= y~x, color='#e08dde', fill='#e08dde') +
  geom_point(alpha=0.5, color="#6c9ebc") +
  geom_label_repel(max.overlaps = 100, cex=0.7) +
  labs(x = "GBs mapped to MAGs", y = "Neutral diversity (effective number of MAGs)") +
  theme_classic() +
  theme(legend.position="none")
```


## Alpha diversity models

### Data preparation for GLMMs

```{r data_prep_models}
diversity.data <- alpha_div %>%
  left_join(sample_metadata, by = join_by(sample == sample)) %>%
  mutate(season=factor(season, levels = c("spring-summer", "autumn", "winter"))) %>%
  right_join(singlem, by = join_by(sample == sample)) %>%
  group_by(species) %>%
  mutate(index500_st = scale(index500, center=T, scale=T)[,1]) %>%
  ungroup() %>%
  filter(sample!="EHI02263") %>% #remove outlier
  filter(development=="Adult") #remove juveniles, nursing and pregnant females

#check whether a low domain-adjusted mapping rate (DAMR) is associated with low diversity estimates 
ggplot(diversity.data, aes(x=est_mapp, y=neutral)) +
  geom_point(size=3, alpha=0.5, color="#6c9ebc") +
  labs(x = "DAMR (mapping rate to MAG catalogue/singleM microbial fraction estimate)", y = "Neutral diversity (effective number of MAGs)") +
  theme_classic() +
  theme(legend.position="none")

diversity.data <- diversity.data %>%
  filter(mags_singlem > 0.8) #remove 5 samples with low DAMR

```

```{r var_distrib, eval=FALSE}

# check y distributions
plot(diversity.data$neutral)
hist(diversity.data$neutral, breaks=30)
d <- density(diversity.data$neutral)
plot(d)

plot(diversity.data$phylo)
hist(diversity.data$phylo, breaks=30)
d <- density(diversity.data$phylo)
plot(d)

plot(diversity.data$func)
hist(diversity.data$func, breaks=30)
d <- density(diversity.data$func)
plot(d)

```



### GLMM - neutral alpha


```{r neutral_model}

neutral.model <-lmer(neutral ~ species + index500 + season + 
            + species:index500 + species:season 
            + (1|animal),
            data=diversity.data, na.action=na.omit,
            control=lmerControl(optimizer="bobyqa",
                                optCtrl=list(maxfun=2e5)))

summary(neutral.model)
Anova(neutral.model, test.statistic='F', type=3) 

#check_model(neutral.model) #not giving the full panel in rmd
resid_panel(neutral.model, smoother = TRUE, qqbands = TRUE)

```
```{r neutral_plots, , fig.width=10}

neutral.model.eff <- allEffects(neutral.model) 
plot(neutral.model.eff)

# plot(effect('species', neutral.model))
# plot(effect('index:500', neutral.model))

```

### GLMM - phylogenetic alpha


```{r phylo_model}

phylo.model<-lmer(phylo ~ species + index500 + season
                    + species:index500 + species:season 
                    + (1|animal),
                    data=diversity.data, na.action=na.omit,
                    control=lmerControl(optimizer="bobyqa",
                                        optCtrl=list(maxfun=2e5)))

summary(phylo.model)
Anova(phylo.model, test.statistic='F', type=3)

resid_panel(phylo.model, smoother = TRUE, qqbands = TRUE)

```


```{r phylo_plots, fig.width=10}

phylo.model.eff <- allEffects(phylo.model) 
plot(phylo.model.eff)
# 
# plot(effect('species', phylo.model))
# plot(effect('area_type', phylo.model))

```

### GLMM - functional (distillr-based) alpha



```{r func2_model2}

func.model<-lmer(func ~ species + index500 + season 
                  + species:index500 + species:season 
                  + (1|animal),
                  data=diversity.data, na.action=na.omit,
                  control=lmerControl(optimizer="bobyqa",
                                      optCtrl=list(maxfun=2e5))) 
  

summary(func.model)
Anova(func.model, test.statistic='F', type=3)

resid_panel(func.model, smoother = TRUE, qqbands = TRUE)

# func.model %>% tbl_regression() %>%
#   add_global_p()

```


```{r func_plots, fig.width=10}

func.model.eff <- allEffects(func.model)
plot(func.model.eff)

```


# Beta diversity metrics

```{r beta_diversity, message=FALSE, eval=TRUE}

#neutral beta div
beta_q1n <-hilldiv2::hillpair(genome_counts, q=1, metric="S")

#phylogenetic beta div (ERRORS)
#beta_q1p <-hilldiv2::hillpair(count_table_cov_size, q=1, tree=tree, metric="S")

#functional beta div
#beta_q1n_fun <-hilldiv2::hillpair(genome_counts_filt, q=1, dist=dist)


```


## Beta Diversity comparisons

### Neutral Beta Diversity

```{r permanova_beta, eval=TRUE}

#neutral beta diversity PERMANOVA

sample_metadata_adonis <- sample_metadata %>%
  filter(sample %in% labels(beta_q1n)) %>%
  arrange(sample) %>%
  #mutate(location=paste0(round(longitude,2),"_",round(latitude,2))) %>%
  select(sample,species,sex,development,area_type,macroarea,season) %>%
  select_if(~ length(unique(.)) > 1) %>% #remove columns with all-identical values
  column_to_rownames(var = "sample") %>%
  as.data.frame()

adonis2(formula=beta_q1n ~ ., data=sample_metadata_adonis[labels(beta_q1n),], permutations=999) %>%
  as.matrix() %>%
  print()

```


```{r nmds_beta_area, eval=TRUE}

beta_q1n_nmds <- beta_q1n %>%
  metaMDS(.,trymax = 500, k=2, verbosity=FALSE) %>%
  vegan::scores() %>%
  as_tibble(., rownames = "sample") %>%
  left_join(sample_metadata, by = join_by(sample == sample))

beta_q1n_nmds %>%
  group_by(species) %>%
  filter(sample !="EHI00420") %>% #remove outliers
  mutate(x_cen = mean(NMDS1, na.rm = TRUE)) %>%
  mutate(y_cen = mean(NMDS2, na.rm = TRUE)) %>%
  ungroup() %>%
  ggplot(., aes(x=NMDS1,y=NMDS2, color=area_type, shape=species, label=sample)) +
  scale_color_manual(values=area_colors) +
  geom_point(size=2.5) + #geom_text(hjust=2, vjust=0) +
  geom_segment(aes(x=x_cen, y=y_cen, xend=NMDS1, yend=NMDS2), alpha=0.2) +
  theme_classic() +
  theme(legend.position="right", legend.box="vertical") +
  guides(color=guide_legend(title="Species"), shape=guide_legend(title="Area type"))
```

```{r nmds_beta_season , eval=TRUE}

beta_q1n_nmds %>%
  group_by(species) %>%
  filter(sample !="EHI00420") %>% #remove outliers
  mutate(x_cen = mean(NMDS1, na.rm = TRUE)) %>%
  mutate(y_cen = mean(NMDS2, na.rm = TRUE)) %>%
  ungroup() %>%
  ggplot(., aes(x=NMDS1,y=NMDS2, color=season, shape=species)) +
  scale_color_manual(values=season_colors) +
  geom_point(size=2.5) +
  geom_segment(aes(x=x_cen, y=y_cen, xend=NMDS1, yend=NMDS2), alpha=0.2) +
  theme_classic() +
  theme(legend.position="right", legend.box="vertical") +
  guides(color=guide_legend(title="Species"), shape=guide_legend(title="Season"))
```

```{r nmds_beta_development}

beta_q1n_nmds %>%
  group_by(species) %>%
  filter(sample !="EHI00420") %>% #remove outliers
  mutate(x_cen = mean(NMDS1, na.rm = TRUE)) %>%
  mutate(y_cen = mean(NMDS2, na.rm = TRUE)) %>%
  ungroup() %>%
  ggplot(., aes(x=NMDS1,y=NMDS2, color=development, shape=species)) +
  geom_point(size=2.5) +
  geom_segment(aes(x=x_cen, y=y_cen, xend=NMDS1, yend=NMDS2), alpha=0.2) +
  theme_classic() +
  theme(legend.position="right", legend.box="vertical") +
  guides(color=guide_legend(title="Species"), shape=guide_legend(title="Development"))
```


### Functional (distillr-based) Beta Diversity

```{r permanova_beta_fun, eval=FALSE}
#functional (distillr-based) beta diversity NMDS
sample_table_adonis_fun <- sample_metadata %>%
  filter(sample %in% labels(beta_q1n_fun$S)) %>%
  arrange(sample) %>%
  #mutate(location=paste0(round(longitude,2),"_",round(latitude,2))) %>%
  select(sample,species,sex,development, area_type,macroarea,season) %>%
  select_if(~ length(unique(.)) > 1) %>% #remove columns with all-identical values
  column_to_rownames(var = "sample") %>%
  as.data.frame()

adonis2(formula=beta_q1n_fun2$S ~ ., data=sample_table_adonis_fun[labels(beta_q1n_fun2$S),], permutations=999) %>%
  as.matrix() %>%
  print()
```

```{r nmds_beta_fun, eval=FALSE}
beta_q1n_fun_nmds <- beta_q1n_fun$S %>%
  metaMDS(.,trymax = 500, k=2, verbosity=FALSE) %>%
  vegan::scores() %>%
  as_tibble(., rownames = "sample") %>%
  left_join(sample_metadata, by = join_by(sample == sample))

beta_q1n_fun_nmds %>%
  group_by(species) %>%
  filter(sample !="EHI00420") %>% #remove outliers
  mutate(x_cen = mean(NMDS1, na.rm = TRUE)) %>%
  mutate(y_cen = mean(NMDS2, na.rm = TRUE)) %>%
  ungroup() %>%
  ggplot(., aes(x=NMDS1,y=NMDS2, color=area_type, shape=species)) +
  scale_color_manual(values=area_colors) +
  geom_point(size=2.5) +
  geom_segment(aes(x=x_cen, y=y_cen, xend=NMDS1, yend=NMDS2), alpha=0.2) +
  theme_classic() +
  theme(legend.position="right", legend.box="vertical") +
  guides(color=guide_legend(title="Area type"), shape=guide_legend(title="Species"))

```



```{r nmds_beta_fun_season, eval=FALSE}
beta_q1n_fun_nmds %>%
  group_by(species) %>%
  filter(sample !="EHI00420") %>% #remove outliers
  mutate(x_cen = mean(NMDS1, na.rm = TRUE)) %>%
  mutate(y_cen = mean(NMDS2, na.rm = TRUE)) %>%
  ungroup() %>%
  ggplot(., aes(x=NMDS1,y=NMDS2, color=season, shape=species)) +
  scale_color_manual(values=squirrel_colors) +
  geom_point(size=2.5) +
  geom_segment(aes(x=x_cen, y=y_cen, xend=NMDS1, yend=NMDS2), alpha=0.2) +
  theme_classic() +
  theme(legend.position="right", legend.box="vertical") +
  guides(color=guide_legend(title="Season"), shape=guide_legend(title="Species"))

```

```{r nmds_beta_fun_development, eval=FALSE}
beta_q1n_fun_nmds %>%
  group_by(species) %>%
  filter(sample !="EHI00420") %>% #remove outliers
  mutate(x_cen = mean(NMDS1, na.rm = TRUE)) %>%
  mutate(y_cen = mean(NMDS2, na.rm = TRUE)) %>%
  ungroup() %>%
  ggplot(., aes(x=NMDS1,y=NMDS2, color=development, shape=species)) +
  geom_point(size=2.5) +
  geom_segment(aes(x=x_cen, y=y_cen, xend=NMDS1, yend=NMDS2), alpha=0.2) +
  theme_classic() +
  theme(legend.position="right", legend.box="vertical") +
  guides(color=guide_legend(title="Development"), shape=guide_legend(title="Species"))

```