
# Diet analysis

```{r load_data09}

rm(list=ls()) #clear environment
load("data/squirrels_data.Rdata")


diet_counts <- read.table("data/diet_counts.tsv",sep = '\t', header = TRUE) 

```

```{r data_preparation09}

red_samples <- sample_metadata %>% 
  filter(species == "Sciurus vulgaris") %>%
  select(sample) %>%
  pull()
grey_samples <- sample_metadata %>% 
  filter(species=="Sciurus carolinensis") %>%
  select(sample) %>% pull()


sample_metadata2 <- read_counts %>%
  pivot_longer(-genome, names_to = "sample", values_to = "value") %>%
  group_by(sample) %>%
  summarise(mags = sum(value)) %>%
  left_join(sample_metadata, by = join_by(sample == sample))  %>%
  mutate(mags_bases = mags*146) %>%
  mutate(lowqual_bases = ((metagenomic_bases+host_bases)/(1-bases_lost_fastp_percent))-(metagenomic_bases+host_bases)) %>%
  mutate(unmapped_bases = metagenomic_bases - mags_bases) %>%
  mutate(unmapped_bases = ifelse(unmapped_bases < 0, 0, unmapped_bases)) %>%
  select(sample, animal, species, sex, development, area_type, index500, season, mags_bases,unmapped_bases,host_bases,lowqual_bases) %>%
  rename(host_sp=species) %>%
  filter(!is.na(host_sp))


plants <- diet_counts %>%
  filter(phylum %in% c("p__Anthophyta", "p__Coniferophyta"),
         !is.na(species),
         species != "") 
  
plants_pivot <- plants %>%
  pivot_longer(-c(1:7), names_to = "sample", values_to = "diet_reads") %>%
  mutate(plant_bases=diet_reads*146) %>%
  left_join(sample_metadata2, by="sample") %>%
  filter(!is.na(host_sp))


# plant_fraction <- plants_pivot %>%
#   group_by(sample) %>%
#   mutate(plant_bases = sum(plant_bases),
#          other_bases = (unmapped_bases - plant_bases),
#          plant_fraction = (plant_bases/unmapped_bases)) %>%
#   select(sample, host_sp, other_bases, plant_bases, unmapped_bases, plant_fraction) %>%
#   distinct(sample, .keep_all = TRUE)
# 
# plant_fraction %>%
#   pivot_longer(-c(sample,host_sp), names_to = "fraction", values_to = "value") %>%
#   filter(fraction != c('unmapped_bases', 'plant_fraction')) %>%
#   mutate(value = value / 1000000000) %>%
#   ggplot(., aes(x = sample, y = value, fill=fraction)) +
#   geom_bar(position="stack", stat = "identity") +
#   scale_fill_manual(values=c("slategrey", "#0B6623")) +
#   labs(x = "Samples", y = "Amount of data (GB)") +
#   theme_classic() +
#   theme(axis.text.x = element_blank())

```

## Overview of plant fraction by species

```{r plant_prev}

plants_pivot %>%
  filter(sample %in% red_samples,
         plant_bases != 0) %>%
  summarise(n = n_distinct(species)) 
plants_pivot %>%
  filter(sample %in% red_samples,
         plant_bases != 0) %>%
  group_by(sample) %>%
  summarise(n = n_distinct(species)) %>%
  summarise(mean = mean(n),
            sd = sd(n))

plants_pivot %>%
  filter(sample %in% grey_samples,
         plant_bases != 0) %>%
  summarise(n = n_distinct(species)) 
plants_pivot %>%
  filter(sample %in% grey_samples,
         plant_bases != 0) %>%
  group_by(sample) %>%
  summarise(n = n_distinct(species)) %>%
  summarise(mean = mean(n),
            sd = sd(n))



prev_red <- plants_pivot %>%
  filter(sample %in% red_samples,
         plant_bases != 0) %>%
  mutate(N = n_distinct(sample)) %>%
  group_by(genus) %>%
  summarise(
    pos = n_distinct(sample),
    N = first(N),
    prevalence = (pos / N)*100) %>%
  mutate(host_sp='Sciurus vulgaris')

#most prevalent plant genera in red
prev_red %>%
  top_n(10, prevalence) %>%
  arrange(desc(prevalence)) %>%
  paged_table()

prev_grey <- plants_pivot %>%
  filter(sample %in% grey_samples,
         plant_bases != 0) %>%
  mutate(N = n_distinct(sample)) %>%
  group_by(genus) %>%
  summarise(
    pos = n_distinct(sample),
    N = first(N),
    prevalence = (pos / N)*100) %>%
   mutate(host_sp='Sciurus carolinensis')
  
#most prevalent plant genera in grey
prev_grey %>%
  top_n(10, prevalence) %>%
  arrange(desc(prevalence)) %>%
  paged_table()


```
```{r plant_prev_plot}


plant_colors <- c("#004c00","#009900","#00cc00","#b2ffb2", 
                  "#8A3324",'#997950', "#EB6841", "#F4AB6A","#FBCEB1", "#43270f", 
                  "#FFC600","#f0dc82","violet","purple", "grey",'#DA70D6' , '#D6CADD')
names(plant_colors) <- c("g__Abies","g__Pinus","g__Picea","g__Taxus",
                         "g__Castanea","g__Corylus","g__Quercus","g__Ulmus","g__Acer","g__Juglans","g__Celtis","g__Crataegus",
                         "g__Zea","g__Aster","other", 'g__Bromus', 'g__Rosaceae_gen_Incertae_sedis' )

top_red <- prev_red %>%
  top_n(10, prevalence)

top_grey <- prev_grey %>%
  top_n(10, prevalence) 

top_genera <- top_red %>%
  full_join(top_grey) %>%
  filter(prevalence>=5) %>%
  distinct(genus) %>%
  pull(genus)


pie_red <- prev_red %>%
  mutate(genus = if_else(genus %in% top_genera, genus, 'other')) %>%
  mutate(genus = factor(genus, levels=c("g__Abies","g__Pinus","g__Picea","g__Taxus",
                         "g__Juglans","g__Castanea","g__Corylus","g__Quercus","g__Ulmus","g__Acer","g__Celtis","g__Crataegus",
                         "g__Aster","g__Zea","other"))) %>%
  ggplot(aes(x="", y=prevalence, fill=genus)) +
  geom_bar(stat="identity", width=1, color="white") +
  coord_polar("y", start=0) +
  scale_fill_manual(values=plant_colors, name="Plant genus", drop=FALSE) +
  theme_void()

pie_grey <- prev_grey %>%
  mutate(genus = if_else(genus %in% top_genera, genus, 'other')) %>%
  mutate(genus = factor(genus, levels=c("g__Abies","g__Pinus","g__Picea","g__Taxus",
                         "g__Juglans","g__Castanea","g__Corylus","g__Quercus","g__Ulmus","g__Acer","g__Celtis","g__Crataegus",
                         "g__Aster","g__Zea","other"))) %>%
  ggplot(aes(x="", y=prevalence, fill=genus)) +
  geom_bar(stat="identity", width=1, color="white") +
  coord_polar("y", start=0) +
  scale_fill_manual(values=plant_colors, name="Plant genus", drop=FALSE) +
  theme_void()
  
ggarrange(pie_red, pie_grey, legend="right", 
          ncol = 2, nrow=1)

prev_both <- prev_red %>%
  rbind(prev_grey)

prev_both %>%
  mutate(genus = if_else(genus %in% top_genera, genus, 'other')) %>%
  mutate(genus = factor(genus, levels=c("g__Abies","g__Pinus","g__Picea","g__Taxus",
                         "g__Juglans","g__Castanea","g__Corylus","g__Quercus","g__Ulmus","g__Acer","g__Celtis","g__Crataegus",
                         "g__Aster","g__Zea","other"))) %>%
  ggplot(aes(x="", y=prevalence, fill=genus, group=host_sp)) +
  geom_bar(stat="identity", width=1, color='white') +
  coord_polar("y", start=0) +
  facet_wrap(~host_sp, scales='free') +
  scale_fill_manual(values=plant_colors, name="Plant genus", drop=FALSE) +
  theme_void()


# plants_pivot %>%
#   #left_join(plant_fraction %>% select(plant_fraction), by='sample') %>%
#   mutate(adj=plant_bases/unmapped_bases) %>%
#   group_by(host_sp) %>%
#   mutate(tot_plant=sum(adj)) %>%
#   group_by(host_sp, genus) %>%
#   summarise(tot_genus=sum(adj),
#          genus_prop=tot_genus/first(tot_plant)) %>%
#   arrange(-genus_prop)

plants_relabun <- plants_pivot %>%
  group_by(host_sp) %>%
  #left_join(plant_fraction %>% select(plant_fraction), by='sample') %>%
  mutate(tot_plant=sum(plant_bases)) %>%
  group_by(host_sp,genus) %>%
  summarise(tot_genus=sum(plant_bases),
            relabun=tot_genus/first(tot_plant)) %>%
  arrange(-relabun) 

plants_relabun %>%
  paged_table()

  plants_relabun %>% mutate(genus = if_else(genus %in% names(plant_colors), genus, 'other')) %>%
  mutate(genus = factor(genus, levels=c("g__Abies","g__Pinus","g__Picea","g__Taxus",
                         "g__Juglans","g__Castanea","g__Corylus","g__Quercus","g__Ulmus","g__Acer","g__Celtis","g__Crataegus",
                         "g__Aster","g__Zea", 'g__Bromus', 'g__Rosaceae_gen_Incertae_sedis',"other"))) %>%
  ggplot(aes(x="", y=relabun, fill=genus, group=host_sp)) +
  geom_bar(stat="identity", width=1, color="white") +
  coord_polar("y", start=0) +
  facet_wrap(~host_sp) +
  scale_fill_manual(values=plant_colors, name="Plant genus", drop=FALSE) +
  theme_void()

```


```{r plant_compo_plot}

plants_by_sample <- plants_pivot %>%
  group_by(sample) %>%
  mutate(total_plant = (sum(plant_bases))) %>%
  filter(plant_bases>0) %>%
  group_by(sample, genus) %>%
  mutate(relabun = plant_bases/first(total_plant))

# plant_colors <- c("#004c00","#009900","#00cc00","#b2ffb2", 
#                   "#8A3324","#CD7F32", "#EB6841", "#F4AB6A","#FFC600", "#C35214", 
#                   "#f0dc82","#FBCEB1","violet","purple", "grey")


# Plot stacked barplot

plants_by_sample %>%
  select(sample,host_sp,genus,relabun) %>%
  mutate(genus = if_else(genus %in% top_genera, genus, 'other')) %>%
  ggplot(aes(x=sample,y=relabun,fill=genus, group=genus))+ 
  geom_bar(stat="identity", colour="white", linewidth=0.02)+ #plot stacked bars with white borders
  scale_fill_manual(values=plant_colors, name="Plant genus", breaks=c("g__Abies","g__Pinus","g__Picea","g__Taxus",
                         "g__Juglans","g__Castanea","g__Corylus","g__Quercus","g__Ulmus","g__Acer","g__Celtis","g__Crataegus",
                         "g__Aster","g__Zea","other")) +
  labs(y = "Relative abundance") +
  guides(fill = guide_legend(ncol = 1)) +
  facet_nested(~host_sp, scales="free", space="free") +
  theme(axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        panel.background = element_blank(),
        panel.border = element_blank(),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.line = element_line(linewidth = 0.5, linetype = "solid", colour = "black"),
        legend.position="right",
        )

```

## Diet diversity estimates

### Rarefaction curves

```{r plant_rarefaction, results='hide'}


genus_matrix <- plants %>%
  select(genus, starts_with("EHI")) %>%  
  group_by(genus) %>%
  summarise(across(.cols = where(is.numeric),.fns = sum)) %>%
  column_to_rownames(var = "genus") 

genus_matrix_filt <- genus_matrix[, which(colSums(genus_matrix) != 0)]
 



# any(!genus_pa %in% c(0, 1))
# which(!genus_pa %in% c(0, 1), arr.ind = TRUE)
  
  
red_counts <- genus_matrix %>%
  select(any_of(red_samples)) %>%
  as.matrix()
grey_counts <- genus_matrix %>%
  select(any_of(grey_samples)) %>%
  as.matrix()
genus_inext <- list(red_counts, grey_counts)

# # Sum counts by species
# red_sum <- genus_matrix %>%
#   select(any_of(red_samples)) %>%
#   rowSums() %>%
#   as.matrix()
# grey_sum <- genus_matrix %>%
#   select(any_of(grey_samples)) %>%
#   rowSums() %>%
#   as.matrix()

# genus_inext <- red_sum %>%
#   cbind(grey_sum) %>%
#   as.data.frame() %>%
#   rename(`Sciurus vulgaris` = V1,
#          `Sciurus carolinensis` = V2)



library(iNEXT)
#DataInfo(genus_inext)
genus_div <- iNEXT(genus_inext, q = c(0,1,2), datatype="abundance", endpoint=300)
genus_div    


red_pa <- genus_matrix %>%
  select(any_of(red_samples)) %>%
  as.matrix()
red_pa[red_pa > 0] <- 1 # convert counts > 0 to 1 (presence/absence)

grey_pa <- genus_matrix %>%
  select(any_of(grey_samples)) %>%
  as.matrix()
grey_pa[grey_pa > 0] <- 1 # convert counts > 0 to 1 (presence/absence)

genus_pa <- list(red_pa,grey_pa)

genus_div2 <- iNEXT(genus_pa, q = 0, datatype="incidence_raw", endpoint=150)
genus_div2    


```
```{r rarefaction_plots}

ggiNEXT(genus_div, type=1, facet.var="Assemblage")

ggiNEXT(genus_div2, type=1)

```

## Alpha diversity

```{r alpha_diet}
#ALPHA DIVERSITY
q0n <- hilldiv2::hilldiv(genus_matrix_filt,q=0) %>% c()
q1n <- hilldiv2::hilldiv(genus_matrix_filt,q=1) %>% c()

# Merge all diversity metrics
diet_div <- cbind(sample=colnames(genus_matrix_filt),richness=q0n,neutral=round(q1n,3)) %>%
  as.data.frame() %>%
  left_join(sample_metadata %>% select(sample,species, animal), by='sample') %>%
  filter(!is.na(species))

# diet_div %>%
#   ggboxplot(x = "species", y = "neutral", group='species', color = "species", fill="white", add="jitter", alpha=0.5) +
#   scale_color_manual(values=squirrel_colors) +
#   scale_fill_manual(values=paste0(squirrel_colors)) +
#   stat_compare_means() +
#   theme_classic() +
#   labs(y = "Neutral Hill numbers") +
#   theme(legend.position = "none",
#         axis.title.x = element_blank())

diet_div %>%
  ggplot(aes(x = species, y = as.numeric(neutral), group=species, color = species, alpha=0.5)) +
  geom_boxplot(fill="white") +
  geom_jitter() +
  scale_color_manual(values=squirrel_colors) +
  scale_fill_manual(values=paste0(squirrel_colors)) +
  stat_compare_means() +
  theme_classic() +
  labs(y = "Richness") +
  theme(legend.position = "none",
        axis.title.x = element_blank())
```


```{r beta_diet}

# beta_plant <-hilldiv2::hillpair(genus_matrix_filt, q=1, metric="S")
# 
# #NMDS
# set.seed(123)
# diet_nmds <- beta_plant %>%
#   metaMDS(.,trymax = 500, k=2, verbosity=FALSE) %>%
#   vegan::scores() %>%
#   as_tibble(., rownames = "sample") %>%
#   left_join(sample_metadata, by = join_by(sample == sample)) 
# 
# # %>%
# #   filter(sample !="EHI00420")  #remove outlier
# 
# 
# diet_nmds %>% group_by(species) %>%
#   mutate(x_cen = mean(NMDS1, na.rm = TRUE)) %>%
#   mutate(y_cen = mean(NMDS2, na.rm = TRUE)) %>%
#   ungroup() %>%
#   ggplot(., aes(x = NMDS1, y = NMDS2)) + 
#   scale_colour_manual(values = squirrel_colors) +
#   scale_fill_manual(values = squirrel_colors) +
#   geom_point(size = 3, alpha=0.8, aes(color = species)) + 
#   stat_ellipse(data=diet_nmds, aes(x = NMDS1, y = NMDS2, group = species, fill=species, color=species),geom="polygon",level=0.95,alpha=0, size=0.8) + 
#   geom_segment(aes(x=x_cen, y=y_cen, xend=NMDS1, yend=NMDS2, color=species), alpha=0.3) +
#   labs(x = "NMDS1", y = "NMDS2") +
#   theme_classic() +
#   theme(axis.text = element_text(colour = "black", size = 12),
#         axis.title = element_text(size = 14, colour = "black"), 
#         legend.text = element_text(size = 11, colour ="black"), 
#         legend.position = "right", axis.title.y = element_text(size = 14), 
#         legend.title = element_text(size = 14, colour = "black"),
#         legend.key=element_blank(),
#         panel.background = element_blank(), 
#         panel.border = element_rect(colour = "black", fill = NA, size = 0.5))
# 


```



## MAGs associated with tannin detoxification

```{r load_annotations}

genome_annotations <- read_tsv("data/genome_annotations.tsv.xz") %>%
    rename(gene=1, genome=2, contig=3)

# genome_kegg <- genome_annotations %>% 
#   group_by(genome) %>%
#   summarise(kegg_id_list = list(unique(kegg_id[kegg_id != "NA"])))
# 
# # genome_counts <- genome_counts %>%
# #   rownames_to_column(., var="genome")
# 
# community_kegg <-genome_kegg %>% 
#   inner_join(genome_counts,by="genome") %>% 
#   pivot_longer(cols = starts_with("EHI"), names_to = "sample", values_to = "abundance") %>%
#   filter(abundance != 0) %>%
#   select(-abundance) %>% 
#   unnest(kegg_id_list) %>% 
#   group_by(sample) %>%
#   summarise(unique_kegg_ids = list(unique(kegg_id_list)))
# 
# species_kegg <- community_kegg %>% 
#   inner_join(sample_metadata,by="sample") %>%
#   select(sample,unique_kegg_ids,species) %>% 
#   unnest(unique_kegg_ids) %>% 
#   group_by(species, unique_kegg_ids) %>%
#   summarise(count = n(), .groups = "drop")

```

```{r tannase}

annot_tannase <- genome_annotations %>%
  filter(str_detect(pfam_hits,"Tannase")) %>%
  left_join(genome_metadata, by="genome")

mags_tannase <- annot_tannase %>%
  pull(genome)

```



```{r tannase_sp}

# genome_counts <- genome_counts %>%
#   rownames_to_column(var="genome")

genome_counts_pivot <- genome_counts %>%
  mutate_at(vars(-genome),~./sum(.)) %>% #apply TSS normalisation
  pivot_longer(-genome, names_to = "sample", values_to = "count") %>% #reduce to minimum number of columns
  left_join(., genome_metadata, by = join_by(genome == genome)) 
# %>% #append taxonomy
  # mutate(phylum = fct_relevel(phylum, rev(ehi_phylum_colors$phylum))) #sort phyla by taxonomy

genome_counts_by_sample <- sample_metadata %>%
  select(sample,species,animal) %>%
  rename(host_sp=species) %>%
  left_join(genome_counts_pivot,., by=join_by("sample" == "sample")) #%>%

  
# Which host species each genome can be found in
genomes_by_species <- genome_counts_by_sample %>%
  filter(count>0) %>%
  group_by(genome) %>%
  mutate(host = if_else(all(host_sp == "Sciurus vulgaris"), "only red",
                        if_else(all(host_sp == "Sciurus carolinensis"), "only grey", "both"))) %>%
  select(genome, host) %>%
  distinct(genome, .keep_all = TRUE) %>%
  left_join(.,genome_metadata, by='genome')

species_tannase <- genomes_by_species %>%
  filter(genome %in% mags_tannase)

host_tannase <- genome_counts_by_sample %>%
  filter(genome %in% mags_tannase)

host_tannase %>%
  group_by(host_sp, genome) %>%
  summarise(mean=mean(count)) %>%
  arrange(-mean)


#samples with at least one tannase-producing MAG by host species
host_tannase %>%
  group_by(host_sp) %>%
  mutate(N = n_distinct(sample)) %>%
  ungroup() %>%
  filter(count>0) %>%
  group_by(host_sp) %>%
  summarise(
    pos = n_distinct(sample),
    N = first(N),
    prevalence = (pos / N)*100
  ) %>%
  # group_by(host_sp) %>%
  # top_n(5, prevalence) %>%
  arrange(host_sp, desc(prevalence)) %>%
  paged_table()

genomes_by_species %>%
  filter(family=="f__Lachnospiraceae") %>%
  group_by(host) %>%
  summarise(n=n_distinct(genus))


```

```{r}

#most prevalent tannase MAGs by host species
host_tannase %>%
  group_by(host_sp) %>%
  mutate(N = n_distinct(sample)) %>%
  ungroup() %>%
  filter(count>0) %>%
  group_by(host_sp, phylum, family, genus, genome) %>%
  summarise(
    pos = n_distinct(sample),
    N = first(N),
    prevalence = (pos / N)*100
  ) %>%
  group_by(host_sp) %>%
  top_n(5, prevalence) %>%
  arrange(host_sp, desc(prevalence)) %>%
  paged_table()

#most abundant tannase MAGs by host species
host_tannase %>%
  # group_by(host_sp, sample, phylum, family, genus, genome) %>%
  # summarise(sum_abundance = sum(count)) %>%
  group_by(host_sp, phylum, family, genus, genome) %>%
  summarise(mean_abundance=mean(count),
            sd=sd(count)) %>%
  arrange(host_sp, desc(mean_abundance)) %>%  # Sort within each host_sp group by abundance
  group_by(host_sp) %>%
  slice_max(order_by = mean_abundance, n = 5) %>%  # Select the top 5 families per host_sp
  ungroup() %>%
  arrange(host_sp, desc(mean_abundance)) %>%
  paged_table()

tannase_summary <- host_tannase %>%
  group_by(sample,host_sp,phylum,family,genus,genome) %>%
  summarise(relabun=sum(count))

# genus_summary %>%
#     group_by(genus) %>%
#     summarise(mean=mean(relabun, na.rm=TRUE),sd=sd(relabun, na.rm=TRUE)) %>%
#     arrange(-mean) %>%
#     paged_table()

genome_arrange <- tannase_summary %>%
    group_by(genome) %>%
    summarise(mean=sum(relabun)) %>%
    arrange(-mean) %>%
    select(genome) %>%
    pull()

tannase_summary %>%
    # left_join(genome_metadata %>% select(genus,phylum) %>% unique(),by=join_by(genus==genus)) %>%
    # left_join(sample_metadata,by=join_by(sample==sample)) %>%
    # filter(genus %in% genus_arrange[1:20]) %>%
    mutate(genome=factor(genome,levels=rev(genome_arrange))) %>%
    filter(relabun > 0) %>%
    ggplot(aes(x=relabun, y=genome, group=genus, color=phylum)) +
        scale_color_manual(values=custom_colors) +
        geom_jitter(alpha=0.5) + 
        facet_grid(genus~host_sp, scales="free", space="free")+
        theme_minimal() + 
        labs(y="Genome", x="Relative abundance", color="Phylum") + 
    guides(colour = guide_legend(override.aes = list(size=5, alpha=0.9)))


tannase_summary_sp <- host_tannase %>%
  group_by(host_sp) %>%
  mutate(N = n_distinct(sample)) %>%
  ungroup() %>%
  filter(count>0) %>%
  group_by(host_sp, phylum, family, genus, genome) %>%
  summarise(
    pos = n_distinct(sample),
    N = first(N),
    prevalence = (pos / N)*100,
    abun = mean(count)
  ) 

tannase_summary_sp %>%
  ggplot(aes(x=prevalence, y=abun, color=host_sp)) +
  geom_point() +
  scale_color_manual(values=squirrel_colors)

# genomes_by_species %>%
#   filter(family=="f__Lachnospiraceae") %>%
#   group_by(host) %>%
#   summarise(n=n_distinct(genus))
```


```{r}
fungi <- diet_counts %>%
  filter(kingdom %in% c("k__Fungi"),
         !is.na(species),
         species != "") 

fungi_pivot <- fungi %>%
  pivot_longer(-c(1:7), names_to = "sample", values_to = "diet_reads") %>%
  mutate(fungi_bases=diet_reads*146) %>%
  left_join(sample_metadata2, by="sample") 


#most prevalent genera in red
fungi_pivot %>%
  filter(sample %in% red_samples,
         fungi_bases != 0) %>%
  mutate(N = n_distinct(sample)) %>%
  group_by(genus) %>%
  summarise(
    pos = n_distinct(sample),
    N = first(N),
    prevalence = (pos / N)*100
  ) %>%
  top_n(10, prevalence) %>%
  arrange(desc(prevalence)) %>%
  paged_table()


#most prevalent genera in grey
fungi_pivot %>%
  filter(sample %in% grey_samples,
         fungi_bases != 0) %>%
  mutate(N = n_distinct(sample)) %>%
  group_by(genus) %>%
  summarise(
    pos = n_distinct(sample),
    N = first(N),
    prevalence = (pos / N)*100
  ) %>%
  top_n(10, prevalence) %>%
  arrange(desc(prevalence)) %>%
  paged_table()


```