---
title: "old_chunks"
author: "Claudia Romeo"
date: "2024-05-02"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r mags_by_species, fig.height=8, fig.width=10, fig.fullwidth=TRUE}

# Which host species the MAG can be found in
genomes_by_species <- genome_counts_by_host %>%
  filter(count>0) %>%
  group_by(genome) %>%
  mutate(host = if_else(all(host_sp == "Sciurus vulgaris"), "only red",
                        if_else(all(host_sp == "Sciurus carolinensis"), "only grey", "both"))) %>%
  select(genome, host) %>%
  distinct(genome, .keep_all = TRUE) %>%
  left_join(.,genome_metadata, by='genome')

genomes_by_species$host <-factor(genomes_by_species$host, levels = c("both", "only red", "only grey"))


# Add host ring
circular_tree_h <-  circular_tree +
  new_scale_fill() +
  scale_fill_manual(values = c("black", "#ed2939", "#92a0ad")) + #"#cc3333", "#999999"
  geom_fruit(
    data=genomes_by_species,
    geom=geom_tile,
    mapping = aes(y=genome, fill=host),
    offset = 0.05,
    width=0.2)
  

#Plot circular tree
circular_tree_h


```



```{r}

table(genomes_by_species$host)

top5_sp <- genome_counts_by_host %>%
  filter(count>0) %>%
  group_by(host_sp,genome) %>%
  summarise(total_count=sum(count)) %>%
  top_n(10, total_count) %>%
  arrange(by_group=host_sp, total_count) %>%
  left_join(.,genome_metadata, by='genome')

top5_sp <- genome_counts_by_host %>%
  filter(count>0) %>%
  group_by(host_sp,genome) %>%
  summarise(total_count=n()) %>%
  top_n(10, total_count) %>%
  arrange(by_group=host_sp, total_count) %>%
  left_join(.,genome_metadata, by='genome')


```


```{r}
# Which host species the genera can be found in
genera_by_species <- genomes_by_species %>%
  group_by(genus) %>%
  mutate(host = if_else(all(host == "only red"), "only red",
                        if_else(all(host == "only grey"), "only grey", "both"))) %>%
  select(genus, host) %>%
  distinct(genus, .keep_all = TRUE) 

table(genera_by_species$host)
```

```{r}
# Which host species families can be found in
families_by_species <- genomes_by_species %>%
  group_by(family) %>%
  mutate(host = if_else(all(host == "only red"), "only red",
                        if_else(all(host == "only grey"), "only grey", "both"))) %>%
  select(family, host) %>%
  distinct(family, .keep_all = TRUE) 

table(families_by_species$host)
```


```{r}
# Which host species families can be found in
orders_by_species <- genomes_by_species %>%
  group_by(order) %>%
  mutate(host = if_else(all(host == "only red"), "only red",
                        if_else(all(host == "only grey"), "only grey", "both"))) %>%
  select(order, host) %>%
  distinct(order, .keep_all = TRUE) 

table(orders_by_species$host)
```

```{r}
# Which host species families can be found in
classes_by_species <- genomes_by_species %>%
  group_by(class) %>%
  mutate(host = if_else(all(host == "only red"), "only red",
                        if_else(all(host == "only grey"), "only grey", "both"))) %>%
  select(class, host) %>%
  distinct(class, .keep_all = TRUE) 

table(classes_by_species$host)
```

```{r additional_seq}

#additonal sequencing needed
# Define the aimed GBs for host and mapped metagenomic data
mags_bases_aim=2
host_bases_aim=5

sequence_fractions_required <- sequence_fractions %>%
  mutate(mags_bases = round(mags_bases / 1000000000,2)) %>%
  mutate(unmapped_bases = round(unmapped_bases / 1000000000,2)) %>%
  mutate(host_bases = round(host_bases / 1000000000,2)) %>%
  mutate(lowqual_bases = round(lowqual_bases / 1000000000,2)) %>%
  mutate(total_bases = mags_bases+unmapped_bases+host_bases+lowqual_bases) %>%
  mutate(mags_bases_fraction = mags_bases/total_bases) %>%
  mutate(mags_bases_difference = mags_bases_aim - mags_bases) %>%
  mutate(meta_required = round(mags_bases_difference / mags_bases_fraction,2)) %>%
  mutate(meta_required = ifelse(meta_required < 0, 0, meta_required)) %>%
  mutate(host_bases_fraction = host_bases/total_bases) %>%
  mutate(host_bases_difference = host_bases_aim - host_bases) %>%
  mutate(host_required = round(host_bases_difference / host_bases_fraction,2)) %>%
  mutate(host_required = ifelse(host_required < 0, 0, host_required)) %>%
  select(sample,mags_bases,unmapped_bases,host_bases,lowqual_bases,meta_required,host_required)

sequence_fractions_required %>%
  select(sample,meta_required,host_required) %>%
  mutate(meta_required = ifelse(meta_required > 20, 21, meta_required)) %>%
  mutate(host_required = ifelse(host_required > 20, 21, host_required)) %>%
  pivot_longer(!sample, names_to = "requirement", values_to = "value") %>%
  mutate(requirement = factor(requirement, levels = c("host_required","meta_required"))) %>%
  ggplot(., aes(x = value, y = sample, fill=requirement, group=requirement)) +
  geom_bar(position="stack", stat = "identity") +
  scale_fill_manual(values=c("#178a94","#d03161")) +
  facet_wrap(~requirement, scales="free_x") +
  labs(x = "Amount of data (GB)", y = "Samples") +
  geom_vline(xintercept = 20, linetype = "dashed", color = "black") +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, size=6),legend.position = "bottom")
```

## Functional ordination of MAGs (kegg)

```{r func_ordination}
# Generate the tSNE ordination
tSNE_func <- Rtsne(X=kegg_table, dims = 2, check_duplicates = FALSE)

# Plot the ordination
tSNE_func$Y %>%
  as.data.frame() %>%
  mutate(genome=rownames(kegg_table)) %>%
  inner_join(genome_metadata, by="genome") %>%
  rename(tSNE1="V1", tSNE2="V2") %>%
  select(genome,phylum,tSNE1,tSNE2, completeness) %>%
  ggplot(aes(x = tSNE1, y = tSNE2, color = phylum, size=completeness))+
  geom_point(shape=16, alpha=0.7) +
  scale_color_manual(values=colors_alphabetic) +
  theme_minimal() +
  theme(legend.position = "none")
```

```{r tree_legend, eval=FALSE}

# Create legend
phyla_legend <- ehi_phylum_colors %>%
  right_join(genome_metadata, by=join_by(phylum == phylum)) %>%
    arrange(match(genome, genome_tree$tip.label)) %>%
  select(phylum, colors) %>%
    unique() %>%
    mutate(phylum = gsub("p__","",phylum)) %>%
  mutate(phylum = factor(phylum, levels = phylum)) %>%
    ggplot() +
      geom_blank() +
      geom_rect(aes(xmin = 1:(nrow(phyla)) - 0.5, xmax = 1:(nrow(phyla)) + 0.5, ymin = 0.19, ymax = 0.2, fill = phylum)) +
        scale_fill_manual(values=rev(phyla$colors)) +
      geom_text(aes(x = 1:(nrow(phyla)), y = 0.15, label = rev(phylum)), angle = 90, hjust = 0, size = 3) +
      theme_void() +
      theme(legend.position = "none")

# Plot legend
phyla_legend
```

```{r beta_diversity, message=FALSE, eval=TRUE}

#neutral beta div
beta_q1n <-hilldiv2::hillpair(genome_counts, q=1, metric="S")

#phylogenetic beta div (ERRORS)
#beta_q1p <-hilldiv2::hillpair(count_table_cov_size, q=1, tree=tree, metric="S")

functional beta div
beta_q1n_fun <-hilldiv2::hillpair(genome_counts_filt, q=1, dist=dist)


```
### Functional (distillr-based) Beta Diversity

```{r permanova_beta_fun, eval=FALSE}
#functional (distillr-based) beta diversity NMDS
sample_table_adonis_fun <- sample_metadata %>%
  filter(sample %in% labels(beta_q1n_fun$S)) %>%
  arrange(sample) %>%
  #mutate(location=paste0(round(longitude,2),"_",round(latitude,2))) %>%
  select(sample,species,sex,development, area_type,macroarea,season) %>%
  select_if(~ length(unique(.)) > 1) %>% #remove columns with all-identical values
  column_to_rownames(var = "sample") %>%
  as.data.frame()

adonis2(formula=beta_q1n_fun2$S ~ ., data=sample_table_adonis_fun[labels(beta_q1n_fun2$S),], permutations=999) %>%
  as.matrix() %>%
  print()
```

```{r nmds_beta_fun, eval=FALSE}
beta_q1n_fun_nmds <- beta_q1n_fun$S %>%
  metaMDS(.,trymax = 500, k=2, verbosity=FALSE) %>%
  vegan::scores() %>%
  as_tibble(., rownames = "sample") %>%
  left_join(sample_metadata, by = join_by(sample == sample))

beta_q1n_fun_nmds %>%
  group_by(species) %>%
  filter(sample !="EHI00420") %>% #remove outliers
  mutate(x_cen = mean(NMDS1, na.rm = TRUE)) %>%
  mutate(y_cen = mean(NMDS2, na.rm = TRUE)) %>%
  ungroup() %>%
  ggplot(., aes(x=NMDS1,y=NMDS2, color=area_type, shape=species)) +
  scale_color_manual(values=area_colors) +
  geom_point(size=2.5) +
  geom_segment(aes(x=x_cen, y=y_cen, xend=NMDS1, yend=NMDS2), alpha=0.2) +
  theme_classic() +
  theme(legend.position="right", legend.box="vertical") +
  guides(color=guide_legend(title="Area type"), shape=guide_legend(title="Species"))

```



```{r nmds_beta_fun_season, eval=FALSE}
beta_q1n_fun_nmds %>%
  group_by(species) %>%
  filter(sample !="EHI00420") %>% #remove outliers
  mutate(x_cen = mean(NMDS1, na.rm = TRUE)) %>%
  mutate(y_cen = mean(NMDS2, na.rm = TRUE)) %>%
  ungroup() %>%
  ggplot(., aes(x=NMDS1,y=NMDS2, color=season, shape=species)) +
  scale_color_manual(values=squirrel_colors) +
  geom_point(size=2.5) +
  geom_segment(aes(x=x_cen, y=y_cen, xend=NMDS1, yend=NMDS2), alpha=0.2) +
  theme_classic() +
  theme(legend.position="right", legend.box="vertical") +
  guides(color=guide_legend(title="Season"), shape=guide_legend(title="Species"))

```

```{r nmds_beta_fun_development, eval=FALSE}
beta_q1n_fun_nmds %>%
  group_by(species) %>%
  filter(sample !="EHI00420") %>% #remove outliers
  mutate(x_cen = mean(NMDS1, na.rm = TRUE)) %>%
  mutate(y_cen = mean(NMDS2, na.rm = TRUE)) %>%
  ungroup() %>%
  ggplot(., aes(x=NMDS1,y=NMDS2, color=development, shape=species)) +
  geom_point(size=2.5) +
  geom_segment(aes(x=x_cen, y=y_cen, xend=NMDS1, yend=NMDS2), alpha=0.2) +
  theme_classic() +
  theme(legend.position="right", legend.box="vertical") +
  guides(color=guide_legend(title="Development"), shape=guide_legend(title="Species"))

```